src/tokens/tokens.utils.ts

```diff
@@ -115,6 +120,78 @@ export async function getCachedTokenBalances(
   return undefined;
 }
 
+export async function getFullToken(chain: Chain, tokenAddr: Token['address']): Promise<TokenFull | null> {
+  const fullTokenMap = await getFullTokens(chain, [tokenAddr]);
+  return fullTokenMap[tokenAddr] || null;
+}
+
+export async function getFullTokens(chain: Chain, tokensAddr: Token['address'][]): Promise<TokenFullMap> {
+  const cachedTokens = await getCachedTokesInfo(tokensAddr);
+  const cachedTokensAddr = cachedTokens.map((token) => token.address);
+
+  const tokensCacheMissMatch = tokensAddr.filter((addr) => !cachedTokensAddr.includes(addr));
+
+  if (tokensCacheMissMatch.length === 0) return mergeTokensFullData(cachedTokens);
+
+  const sdk = await chain.getSdk();
+  let tokensInfo: Token[] = [];
+  try {
+    const sdkTokensInfo = await sdk.tokens.loadTokens(tokensCacheMissMatch);
+    tokensInfo = Object.values(sdkTokensInfo);
+  } catch (e) {
+    console.warn(`Faild to load tokens from chain node ${e}`);
+  }
+
+  if (tokensInfo.length > 0) await cacheTokensInfo(tokensInfo);
+
+  const tokensList = tokensInfo.concat(cachedTokens);
+
+  return mergeTokensFullData(tokensList);
+}
+
+export async function getCachedTokesInfo(tokensAddr: Token['address'][]): Promise<Token[]> {
+  const mapper = getDataMapper();
+
+  const tokensInfo: Token[] = [];
+
+  for (const addr of tokensAddr) {
+    try {
+      tokensInfo.push(await mapper.get(Object.assign(new TokenInformationSnapshot(), { address: addr })));
+    } catch (e) {
+      console.log(`Token snapshot not found on addr ${addr} `);
+    }
+  }
+
+  return tokensInfo;
+}
+
+export async function cacheTokensInfo(tokens: Token[]): Promise<void> {
+  const mapper = getDataMapper();
+
+  const tokensInfoMeta = tokens.map((token) => Object.assign(new TokenInformationSnapshot(), token));
+
+  try {
+    for await (const persisted of mapper.batchPut(tokensInfoMeta)) {
+      if (!persisted) console.warn('Failed to save token info');
+    }
+  } catch (e) {
+    console.warn(`Failed to save tokens info ${e}`);
+  }
+}
+
+export function mergeTokensFullData(tokens: Token[]): TokenFullMap {
```

Comment:
we could probably remove this. i think protocol tokens we don't need any more because that was just being used to look up totken definitions cross chain using `getToken` i think. 

may want to see how that manifests in other parts of the code particularly `getVaultTokenPrice`

however - i think it doesn't matter because you can just use the chain there to call the same methods you have here (sans mergeTokensFullData) to get the same information

Reply:
you are right, missed that
then also we should remove `addr`, `decimals` and other field from chain tokens, dunno how the build passed